python/dgl/nn/pytorch/conv/cugraph_relgraphconv.py

```diff
@@ -0,0 +1,388 @@
+            agg_fwd_func = agg_hg_basis_post_fwd_int64
+        else:
+            raise TypeError(
+                f"Supported ID type: torch.int32 or torch.int64, "
+                f"but got {g.idtype}"
+            )
+        ctx.idtype = g.idtype
+
+        _in_feat = feat.shape[-1]
+        indptr, indices, edge_ids = g.adj_sparse("csc")
+        # edge_ids is in a mixed order, need to permutate incoming etypes
+        # and stash the result for backward propagation
+        ctx.edge_types_int32 = edge_types[edge_ids.long()].int()
+        # node_types are not needed by the post-variant RGCN aggregation functions
+        _num_node_types = 0
+        _out_node_types = _in_node_types = None
+
+        mfg = mfg_csr_func(
+            fanout,
+            g.dstnodes(),
+            g.srcnodes(),
+            indptr,
+            indices,
+            _num_node_types,
+            num_rels,
+            _out_node_types,
+            _in_node_types,
+            ctx.edge_types_int32,
+        )
+        ctx.mfg = mfg
+
+        if coeff is None:
+            leading_dimension = num_rels * _in_feat
+        else:
+            _num_bases = coeff.shape[-1]
+            leading_dimension = _num_bases * _in_feat
+
+        agg_output = th.empty(
+            g.num_dst_nodes(),
+            leading_dimension,
+            dtype=th.float32,
+            device=ctx.device,
+        )
+        if coeff is None:
+            agg_fwd_func(agg_output, feat.detach(), mfg)
+        else:
+            agg_fwd_func(
+                agg_output,
+                feat.detach(),
+                mfg,
+                weights_combination=coeff.detach()
+            )
+
+        ctx.save_for_backward(feat, coeff)
+        return agg_output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        r"""Compute the backward pass of R-GCN aggregation layer.
+
+        Parameters
+        ----------
+        grad_output : torch.Tensor, dtype=torch.float32
+            Gradient of loss function w.r.t output.
+        """
+        feat, coeff = ctx.saved_tensors
+
+        if ctx.idtype == th.int32:
+            agg_bwd_func = agg_hg_basis_post_bwd_int32
+        else:
+            agg_bwd_func = agg_hg_basis_post_bwd_int64
+
+        grad_feat = th.empty_like(feat, dtype=th.float32, device=ctx.device)
+        if coeff is None:
+            grad_coeff = None
+            agg_bwd_func(grad_feat, grad_output, feat.detach(), ctx.mfg)
+        else:
+            grad_coeff = th.empty_like(
+                coeff, dtype=th.float32, device=ctx.device
+            )
+            agg_bwd_func(
+                grad_feat,
+                grad_output,
+                feat.detach(),
+                ctx.mfg,
+                output_weight_gradient=grad_coeff,
+                weights_combination=coeff.detach()
+            )
+
+        return None, None, None, None, grad_feat, grad_coeff
+
+
+class CuGraphRelGraphConv(nn.Module):
```

Comment:
Can we change from `pytorch/conv/cugraph_relgraphconv.py/CuGraphRelGraphConv` to `pytorch/conv/cugraph/relgraphconv.py/RelGraphConv`? I just find this name a bit tedious. Also we might want to handle all dependency checks for cugraph once in the same subdirectory. @jermainewang

Reply:
What's the confusion here?

Reply:
First, we have dependency checks like 

```
try:
    from pylibcugraphops.legacy.aggregators.node_level import (
        agg_hg_basis_post_bwd_int32,
        agg_hg_basis_post_bwd_int64,
        agg_hg_basis_post_fwd_int32,
        agg_hg_basis_post_fwd_int64,
    )
    from pylibcugraphops.legacy.structure.graph_types import (
        message_flow_graph_hg_csr_int32,
        message_flow_graph_hg_csr_int64,
    )
except ModuleNotFoundError:
    raise ModuleNotFoundError(
        "dgl.nn.CuGraphRelGraphConv requires pylibcugraphops to be installed."
    )
```

If we have a dedicated subdirectory for CuGraph, then it can be done there rather than in individual files.

Second, I think it's better to do

```
from dgl.nn.cugraph import AConv, BConv, ...
```

than

```
from dgl.nn import CuGraphAConv, CuGraphBConv
```

These are minor issues. I'm fine to wait for a few more modules to see what will happen.